src/main/java/de/tum/cit/aet/artemis/atlas/service/learningpath/LearningPathRecommendationService.java

```diff
@@ -145,43 +149,178 @@ public RecommendationState getRecommendedOrderOfAllCompetencies(LearningPath lea
+            LearningPathNavigationObjectDTO previousLearningObject = getPreviousLearningObjectInCompetency(learningPath, currentLearningObject, currentCompetency,
+                    learnObjectsOfCurrentCompetency);
+            if (previousLearningObject != null) {
+                return previousLearningObject;
+            }
+        }
+
+        RecommendationState recommendationStateWithAllCompetencies = getRecommendedOrderOfAllCompetencies(learningPath);
+        int indexToSearch;
+        if (currentLearningObject == null || currentLearningObject.repeatedTest()) {
+            Optional<CourseCompetency> firstNonMasteredCompetency = recommendationStateWithAllCompetencies.recommendedOrderOfCompetencies().stream()
+                    .map(recommendationState.competencyIdMap::get)
+                    .filter(competency -> recommendationState.competencyMastery().get(competency.getId()) >= competency.getMasteryThreshold()).findFirst();
+            CourseCompetency competencyToSearch = firstNonMasteredCompetency
+                    .orElse(recommendationState.competencyIdMap().get(recommendationState.recommendedOrderOfCompetencies().getLast()));
+            indexToSearch = recommendationStateWithAllCompetencies.recommendedOrderOfCompetencies().indexOf(competencyToSearch.getId());
+        }
+        else {
+            indexToSearch = recommendationStateWithAllCompetencies.recommendedOrderOfCompetencies().indexOf(currentLearningObject.competencyId());
+        }
+
+        for (int i = indexToSearch; i >= 0; i--) {
+            CourseCompetency competency = recommendationStateWithAllCompetencies.competencyIdMap()
+                    .get(recommendationStateWithAllCompetencies.recommendedOrderOfCompetencies().get(i));
+
+            boolean repeatedTests = currentLearningObject != null && currentLearningObject.repeatedTest();
+            List<LearningObject> learningObjects = getOrderOfLearningObjectsForCompetency(competency, learningPath.getUser(), repeatedTests);
+            if (learningObjects.isEmpty()) {
+                continue;
+            }
+
+            if (i == indexToSearch && currentLearningObject != null && !currentLearningObject.repeatedTest()) {
+                LearningPathNavigationObjectDTO previousLearningObject = getPreviousLearningObjectInCompetency(learningPath, currentLearningObject, competency, learningObjects);
+                if (previousLearningObject != null) {
+                    return previousLearningObject;
+                }
+            }
+            else {
+                LearningObject lastLearningObject = learningObjects.getLast();
+                return LearningPathNavigationObjectDTO.of(lastLearningObject, false, learningObjectService.isCompletedByUser(lastLearningObject, learningPath.getUser()),
+                        competency.getId());
             }
```

Comment:
I feel like the control flow here is a bit convoluted. Could you maybe add a short comment on what you do here?